src/ekat/kokkos/ekat_kokkos_utils.hpp

```diff
@@ -77,6 +78,60 @@ struct ExeSpaceUtils {
   static TeamPolicy get_team_policy_force_team_size (Int ni, Int team_size) {
     return TeamPolicy(ni, team_size);
   }
+
+  template <typename TeamMember, class Lambda, typename ValueType>
+  static KOKKOS_INLINE_FUNCTION
+  void parallel_reduce (const TeamMember& team,
+			const int& begin,
+			const int& end,
+                        const Lambda& lambda,
+                        const bool serialize,
+                        ValueType& result)
+  {
+    if (serialize)
+    {
+      // We want to get C++ on GPU to match F90 on CPU. Thus, need to
+      // serialize parallel reductions.
+
+      // All threads init result.
+      // NOTE: we *need* an automatic temporary, since we do not know
+      //       where 'result' comes from. If result itself is an automatic
+      //       variable, using it would be fine (only one vector lane would
+      //       actually have a nonzero value after the single). But if
+      //       result is taken from a view, then all vector lanes would
+      //       see the updated value before the vector_reduce call,
+      //       which will cause the final answer to be multiplied by the
+      //       size of the warp.
+      auto local_tmp = Kokkos::reduction_identity<ValueType>::sum();
+
+      Kokkos::single(Kokkos::PerThread(team),[&] {
+      for (int k=begin; k<end; ++k)
+        lambda(k, local_tmp);
+        });
+
+      result = local_tmp;
+    }
+    else
+      Kokkos::parallel_reduce(Kokkos::TeamThreadRange(team, begin, end), lambda, result);
+  }
+
+  template<typename PackType, typename ValueType, bool Serialize=false>
+  static KOKKOS_INLINE_FUNCTION
+  void reduce_add (const PackType& p,
+                   ValueType& result)
+  {
+    if (Serialize)
+    {
+      for (int i = 0; i < PackType::n; ++i)
+        result += p[i];
+    }
+    else
+    {
+      typename PackType::scalar sum = p[0];
+      vector_simd for (int i = 1; i < PackType::n; ++i) sum += p[i];
```

Comment:
It might be interesting to write a micro reproducer for a simd reduction (without ekat, just plain cpp), and check if vector_simd vectorizes. If not, try with something like

```
#define vector_simd_reduce(op,result) _Pragma("omp simd reduction(op:result)")
```
(if openmp is enabled, and expand to nothing otherwise). I suspect `vector_simd` might not always vectorize, while the omp simd reduction will. Or, at least, the latter will vectorize _more_ times.

Reply:
Moved to a new PR